language/diem-vm/mvhashmap/src/lib.rs

```diff
@@ -0,0 +1,235 @@
+            .get(&version)
+            .ok_or(Error::UnexpectedWrite)
+    }
+
+    /// Write to `key` at `version`.
+    /// Function will return an error if the write is not in the initial `possible_writes` list.
+    pub fn write(&self, key: &K, version: Version, data: V) -> Result<(), Error> {
+        // By construction there will only be a single writer, before the
+        // write there will be no readers on the variable.
+        // So it is safe to go ahead and write without any further check.
+        // Then update the flag to enable reads.
+
+        let entry = self.get_entry(key, version)?;
+
+        #[cfg(test)]
+        {
+            // Test the invariant holds
+            if entry.is_assigned() {
+                panic!("Cannot write twice to same entry.");
+            }
+        }
+
+        entry.write(data);
+
+        Ok(())
+    }
+
+    /// Skips writing to `key` at `version` if that entry hasn't been assigned.
+    /// Function will return an error if the write is not in the initial `possible_writes` list.
+    pub fn skip_if_not_set(&self, key: &K, version: Version) -> Result<(), Error> {
+        // We only write or skip once per entry
+        // So it is safe to go ahead and just do it.
+        let entry = self.get_entry(key, version)?;
+
+        // Test the invariant holds
+        if !entry.is_assigned() {
+            entry.skip();
+        }
+
+        Ok(())
+    }
+
+    /// Skips writing to `key` at `version`.
+    /// Function will return an error if the write is not in the initial `possible_writes` list.
+    /// `skip` should only be invoked when `key` at `version` hasn't been assigned.
+    pub fn skip(&self, key: &K, version: Version) -> Result<(), Error> {
+        // We only write or skip once per entry
+        // So it is safe to go ahead and just do it.
+        let entry = self.get_entry(key, version)?;
+
+        #[cfg(test)]
+        {
+            // Test the invariant holds
+            if entry.is_assigned() {
+                panic!("Cannot write twice to same entry.");
+            }
+        }
+
+        entry.skip();
+        Ok(())
+    }
+
+    /// Get the value of `key` at `version`.
+    /// Returns Ok(val) if such key is already assigned by previous transactions.
+    /// Returns Err(None) if `version` is smaller than the write of all previous versions.
+    /// Returns Err(Some(version)) if such key is dependent on the `version`-th transaction.
+    pub fn read(&self, key: &K, version: Version) -> Result<&V, Option<Version>> {
+        let tree = self.data.get(key).ok_or(None)?;
+
+        let mut iter = tree.range(0..version);
+
+        while let Some((entry_key, entry_val)) = iter.next_back() {
+            if *entry_key < version {
+                match entry_val.get() {
+                    // Entry not yet computed, return the version that blocked this query.
+                    None => return Err(Some(*entry_key)),
+                    Some(None) => continue,
```

Comment:
this is skip, right? maybe add a comment at the end of the line? 
=> continue, // entry skipped